### PR TITLE
*: upgrade Go to 1.26

### LINKED ISSUE #11193: *: upgrade Go to 1.26 and update static analysis tools
## Enhancement Task

Upgrade active Go version declarations and toolchain references from the Go 1.25 release line to Go 1.26, using the latest Go 1.26 patch release for module directives, and update the static analysis toolchain for Go 1.26 compatibility.

- [ ] Update all Go module directives from Go 1.25.12 to Go 1.26.8.
- [ ] Update GitHub Actions, Docker, build prerequisites, and documentation references from Go 1.25 to Go 1.26 without pinning a patch version where the previous reference did not pin one.
- [ ] Upgrade golangci-lint from v2.6.0 to the latest compatible release (currently v2.13.2).
- [ ] Adapt `.golangci.yml` and the codebase to new checks without changing existing runtime semantics or application logic.
- [ ] Review other Go-dependent static analysis and build tools and update only where required.
- [ ] Run module tidy, build, static analysis, and relevant tests with Go 1.26.8.

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

PD uses Go 1.25 and golangci-lint v2.6.0. Updating the toolchain also exposes a resource-group test that dereferences a nil error while deletion is still propagating, GC tests that leave tuning active, and HTTP clients that rely on redirects when an endpoint ends in a slash.

Issue Number: Close #11193

### What is changed and how does it work?

```commit-message
Upgrade the CI and Docker Go release line to 1.26, set the four modules' minimum Go version to 1.26.8, and pin the repository-local golangci-lint binary to v2.13.2.

Adapt lint diagnostics while preserving TSO batch ordering and capacity, hot-peer iteration, serialization receiver contracts, TLS verification, proxy forwarding, and Pyroscope authentication. Share checked key-range construction between labeler helpers and tests, retaining the public wrapper, nil output, and odd-input panic contract; the panic text changes.

Normalize only the joining slashes in pd-ctl and PD HTTP client URLs, preserving escaped paths and query strings. Configuration writes now reach the intended endpoint directly instead of depending on redirect status codes to preserve the request method and body. Check both request construction and the resulting PD configuration.

Make resource-group deletion assertions tolerate the still-present state before the asynchronous watch arrives, and check that state before deletion. Give the memory-tuner test its own finalizer, drain its reset work, and restore process memory settings; stop the GOGC test's tuner and release its retained heap. Serialize finalizer callbacks with stop so an already-started callback cannot overwrite restored settings after stop returns. The memory-pressure thresholds and GC tuning calculations stay unchanged.

Check scheduler diagnostic readiness using scheduling eligibility and a fresh result. A pending result can also reflect store or region constraints, so it must not prevent testing the live transition to a zero scheduling limit.

Strengthen store-region response assertions and cover strict URL parsing boundaries. Keep the original single-range keyspace builder to avoid an unnecessary refactor.
```

### Check List

Tests

- Unit test
- Integration test
- Manual test
  - `GOTOOLCHAIN=go1.26.8 make check` and `make build`.
  - Failpoint-aware race/deadlock tests, including both resource-manager deployment modes and the PD HTTP client/configuration paths.
  - Before/after and ablation checks for URL construction, resource-group waiting, and finalizer callback completion; regression tests for deterministic TSO overflow handling.
  - Fixed-scale region-heartbeat workload with GOGC tuning, with and without a memory limit, comparing Go 1.25.12, Go 1.26.8, and Go 1.26.8 with Green Tea disabled. This is a bounded workload comparison, not a production benchmark.
  - Full client basic tests use an isolated network namespace because an existing test requires loopback port 80 to be unused.

Code changes

- Has the configuration change

### Release note

```release-note
Upgrade the required Go toolchain to 1.26.8 and fix trailing-slash endpoint handling in pd-ctl and the PD HTTP client. Stopping a GC tuner now waits for any in-flight finalizer callback to finish.

Go's HTTP path normalization now uses 307 redirects: clients following redirects can execute noncanonical write requests that previously became GET requests. Use canonical endpoint paths. Malformed multi-colon URL hosts are rejected earlier; bracket IPv6 addresses, including in dashboard-address. TLS inherits Go 1.26's additional hybrid key-exchange groups, and PD exports additional Go GC/scheduler metrics.
```

### CODE REVIEW

File: pkg/utils/tsoutil/tso_dispatcher.go [hunk 1]
```diff
@@ -144,6 +144,7 @@ func (s *TSODispatcher) dispatch(
 			pendingTSOReqCount := len(tsoQueue.requestCh) + 1
 			requests[0] = first
 			for i := 1; i < pendingTSOReqCount; i++ {
+				// #nosec G602 -- len(requests) equals cap(requestCh)+1, which bounds pendingTSOReqCount.
```
Comment (rleungx) [inline pkg/utils/tsoutil/tso_dispatcher.go]: Shall we fix them instead of ignoring them?

Comment (JmPotato) [reply pkg/utils/tsoutil/tso_dispatcher.go]: Addressed this G602 finding in 5d88bfcef5 by appending into the reused, preallocated batch slice and removing the suppression. The queue-length snapshot stays at the same point, so batch size and request order are preserved. Capacity remains maxMergeRequests+2, which covers the first request plus a full channel; append therefore does not grow the backing array. The existing tsoutil tests pass with race/deadlock enabled, and make check passes with golangci-lint v2.13.2.

**Review by bufferflies — APPROVED** [review]

File: tests/server/api/region_label_test.go [hunk 2]
```diff
@@ -118,6 +118,7 @@ func (suite *regionLabelTestSuite) checkGetSet(cluster *tests.TestCluster) {
 func makeKeyRanges(keys ...string) []any {
 	var res []any
 	for i := 0; i < len(keys); i += 2 {
+		// #nosec G602 -- Test inputs are start/end pairs; odd input intentionally panics.
```
Comment (rleungx) [inline tests/server/api/region_label_test.go]: Do we also need to fix them?

Comment (JmPotato) [reply tests/server/api/region_label_test.go]: Revised.

File: tests/server/api/region_test.go [hunk 3]
```diff
@@ -523,6 +523,7 @@ func (suite *regionTestSuite) checkStoreRegions(cluster *tests.TestCluster) {
 	re.NoError(err)
 	re.Len(regionIDs, r5.Count)
 	for i, r := range r5.Regions {
+		// #nosec G602 -- Preserve the existing assertion loop over the returned regions.
```
Comment (rleungx) [inline tests/server/api/region_test.go]: How about G602?

Comment (JmPotato) [reply tests/server/api/region_test.go]: Addressed in 83b2325fe7. The G602 in `checkStoreRegions` is now addressed by asserting `Count == 1` and `len(Regions) == 1` before comparing the sole ID with 4. This intentionally strengthens the test: the previous code accepted `Count == 1` with an empty array, and panicked for extra matching entries. No server behavior is changed.

The other key-range G602 is addressed in a shared `keyutil.BuildKeyRangeMaps` helper. It converts each pair to `[2]string` before accessing its elements. The two private test helpers are removed, and the labeler and keyspace builders reuse the same implementation. `labeler.MakeKeyRanges` remains as a compatibility entry point. Empty input still returns nil, pair order and values are unchanged, and odd input still panics (the runtime panic text changes).

I also reviewed the remaining suppressions. Removing all of them would require changes beyond behavior-preserving Go/tooling adaptation:

- `checkRules` / G602: the existing `require.Len` already guarantees matching lengths before comparison. Keeping the per-element assertions preserves the useful expected/actual diagnostics; replacing them with a boolean slice comparison offers little benefit here.
- TLS / G123: `VerifyConnection` also runs on resumed connections, unlike the existing callback. Migrating would change which connections are accepted; explicitly skipping resumptions would merely preserve the issue behind a different callback.
- Dashboard `Director` / SA1019: `Rewrite` changes forwarding-header handling, hop-by-hop processing order and query normalization. A direct substitution is not equivalent.
- Contexts / G118: `OrderedSingleFlight` already stores and invokes cancel on cancellation/completion. The router and benchmark findings concern existing lifecycle handling; adding a constructor-level `defer cancel()` or deleting the child contexts changes that lifecycle.
- Dynamic URLs / G704 and G710: proxying to discovered/configured endpoints, dashboard redirects and the test harness URL are existing functionality. Adding target/scheme restrictions or changing URL normalization would change the supported behavior.
- Pyroscope / SA1019: the pinned v1.2.0 client requires the `AuthToken` field during initialization for its legacy cloud endpoint. Replacing it with an Authorization header would fail that check and can change authentication precedence.
- pd-ut / G702 and G706: executing the selected test binary and passing Go arguments are intentional runner functionality; escaping compiler diagnostics changes log output. Merely switching to calls the analyzer does not recognize would not fix the underlying concerns.
- `recvcheck`: the mixed serialization receivers preserve method sets and mutation semantics. Unifying their receivers would change interface satisfaction or deserialization behavior.

These local exceptions preserve the current scope; they are not a claim that all pre-existing findings are harmless. Behavioral/security hardening should be reviewed separately rather than hidden inside the toolchain upgrade.

Validation: `make check` passed with Go 1.26.8 and golangci-lint v2.13.2 (reusing the generated dashboard assets). The keyutil/labeler suites and relevant keyspace, merge/split checker, region API and pd-ctl tests passed with race/deadlock enabled in both normal and nextgen configurations.

**Review by rleungx — APPROVED** [review]

**Comment by SongXinbai** [comment]
Heads-up: moving the main module's `go` directive to `1.26.8` enables Go 1.26's `urlstrictcolons` behaviour, which breaks `--join` when it is given multiple endpoints.

`Config.Adjust` validates the whole comma-separated list with a single `url.Parse` ([server/config/config.go#L394-L398](https://github.com/tikv/pd/blob/master/server/config/config.go#L394-L398)), so on this branch:

```go
cfg := NewConfig()
cfg.Join = "http://pd-0:2379,http://pd-1:2379"
err := cfg.Adjust(nil, false)
// failed to parse join addr:http://pd-0:2379,http://pd-1:2379,
// err:parse "http://pd-0:2379,http://pd-1:2379": invalid port ":2379,http:" after host
```

PD then fails during configuration validation, before contacting any existing member, so the process never starts. This is a documented configuration: the PD config docs state that multiple advertise client URLs are separated by commas, and TiDB Operator's PD recovery workflow generates `--join=http://demo-pd-0.demo-pd-peer.demo.svc:2380,http://demo-pd-1.demo-pd-peer.demo.svc:2380`.

CI does not catch this because no test exercises a multi-endpoint `--join`, so `./server/config` passes on this branch.

Filed as #11202, with a fix in #11203 that uses the existing `parseUrls` helper. Flagging it here only for visibility — no action needed on this PR.

**Comment by JmPotato** [comment]
/hold

Wait for https://github.com/tikv/pd/pull/11203 to be merged first.

**Review by rleungx — APPROVED** [review]

File: pkg/gctuner/finalizer.go [hunk 4]
```diff
@@ -16,24 +16,27 @@ package gctuner
...
+	"sync"
 )
 
 type finalizerCallback func()
 
 type finalizer struct {
 	ref      *finalizerRef
 	callback finalizerCallback
-	stopped  atomic.Int32
+	mu       sync.Mutex
```
Comment (niubell) [inline pkg/gctuner/finalizer.go:27]: What's the concern with changing atomic int32 to mutex?

**Review by niubell — APPROVED** [review]
LGTM
